### ISSUE TITLE
as_list() is not defined on an unknown TensorShape.

### ISSUE THREAD

**Issue by Xiaocong6** [google-research/morph-net#45]
Hi, I got a problem when I used the VGG. The code is as follows:
```
def vgg_a(inputs,num_classes=10,is_training=True,dropout_keep_prob=0.5,spatial_squeeze=True,
          scope='vgg_a',fc_conv_padding='SAME',global_pool=False):
          .........
            net = slim.conv2d(net,512,[3,3],scope='conv5_2')
            net = slim.max_pool2d(net,[2,2],scope='pool5')
            net = slim.conv2d(net,4096,[1,1],padding=fc_conv_padding,scope='fc6')
            net = slim.dropout(net,dropout_keep_prob,is_training=is_training,scope='dropout6')
            net = slim.conv2d(net,4096,[1,1],scope='fc7')
            if num_classes:
                net = slim.dropout(net,dropout_keep_prob,is_training=is_training,scope='dropout7')
                net = slim.conv2d(net,num_classes,[1,1],activation_fn=None,normalizer_fn=None,scope='fc8')
    return net

if __name__ == '__main__':
    train_x,train_y,test_x,test_y = prepare_data()

    x = tf.placeholder(tf.float32,[None,image_size,image_size,3])
    y_ = tf.placeholder(tf.float32,[None,class_num])
    dropout_keep_prob = tf.placeholder(tf.float32)
    learning_rate = tf.placeholder(tf.float32)
    is_training = tf.placeholder(tf.bool)

    logits,_ = vgg_a(inputs=x, dropout_keep_prob=dropout_keep_prob)

    network_regularizer = flop_regularizer.GroupLassoFlopsRegularizer([logits.op], threshold=1e-3)
    regularization_strength = 1e-7
    regularizer_loss = (network_regularizer.get_regularization_term() * regularization_strength)
    
    .........
   _,batch_loss = sess.run([train_step,cross_entropy],feed_dict={
                    x:batch_x, y_:batch_y, dropout_keep_prob:dropout_rate, learning_rate:lr,
                    is_training: True})
   batch_acc = accuracy.eval(feed_dict={x:batch_x, y_:batch_y, dropout_keep_prob: 1.0,
                                                     is_training: True})
```
![无标题](https://user-images.githubusercontent.com/44869509/56958647-6b7dc780-6b7d-11e9-829b-5176b7e91ed5.png)

   I would like to ask what is the cause of this error, does not support dropout? “logits,_ = vgg_a(inputs=x)” will not be wrong,  but dropout requires different parameters in the train and accuracy eval. Is there any solution?
Thanks!

**Comment by pkch** [comment]
IIUC this is because you leave batch size as `None` when building the graph. While normally it's fine, MorphNet framework expects to see the full shape of the ops it regularizes. Could you try including the actual batch size during the graph construction?

**Comment by Xiaocong6** [comment]
Maybe this is not caused by the batch size being None, 
x = tf.placeholder(tf.float32,[250,image_size,image_size,3]) will still go wrong.
I looked at op.outputs[0].shape during the debugging process, as shown below. It can be seen that there is an error in the dropout layer, the shape is unknown. The regularizer I used is flop_regularizer.GroupLassoFlopsRegularizer, flop_regularizer.GammaFlopsRegularizer will not be wrong here. What is the cause of this, what special treatment is needed for dropout?
Thanks for your answer!
![error](https://user-images.githubusercontent.com/44869509/57191476-cd14ac00-6f58-11e9-8465-2c648144e0cc.png)

**Comment by eladeban** [comment]
Hi Xiaocong6,

I have just submitted [PR#51](https://github.com/google-research/morph-net/pull/51) which unifies the way the GammaFlopsRegularizer and GroupLassoFlopsRegularizer are created.
I hope this will solve the problem. 

Let me know.

**Comment by Xiaocong6** [comment]
Sorry, I want to know what I should do. Copy op_handlers.py and op_handlers_test.py into the morph_net/framework or re-download and python setup.py install? But this didn't work. What should I do to make use of your changes?

**Comment by pkch** [comment]
From your existing git repo, you can run `git pull` to update to the current master; redownloading and installing from scratch should also work.

If you are still seeing this error, could you copy a snippet of code that we can run to reproduce the issue?  We cannot run the code you copied at the beginning of this issue because it contains references to functions not defined in it (e.g., `prepare_data`, etc.).

**Comment by Xiaocong6** [comment]
Thanks, I will try it soon and tell you the results.

**Comment by eladeban** [comment]
@Xiaocong6,  closing. let us know if things worked out for you eventually.
